Feature/ignore nonmsgpack

erikleitch opened the pull request:
This branch includes two functional changes, and one change to a unit test file:
- workitems.cc has been modified so that records that are non-msgpack encoded will be ignored during a streaming fold.  This handles the case where msgpack-encoded and non-magpack-encoded records are interleaved in the same leveldb file
- eleveldb:do_streaming_fold() has been modified to log errors thrown up from the C++ layer during filtering
- I've added test function r() to test/sut.erl as another example of decoding TS values returned from eleveldb, and of correct filter formatting

Commit:
Added code to ignore non-decodable data

Commit:
Added logging of error messages in eleveldb:do_streaming_fold() and added another example of reading TS values to sut.erl

Commit:
Elaborated some error messages

hmmr commented on c_src/extractor.cc:
```diff
@@ -338,6 +338,18 @@ DataType::Type Extractor::cTypeOf(ErlNifEnv* env, ERL_NIF_TERM tuple, bool throw
+{
+    const char* ptr = data;
+    ptr++;
+    unsigned char vers     = (*ptr++);
```
There is a magic number at offset 0: https://github.com/basho/riak_kv/blob/develop/src/riak_object.erl#L79. Perhaps you should be checking it as well?

hmmr commented on c_src/workitems.cc:
```diff
@@ -767,17 +769,50 @@ work_result RangeScanTask::operator()()
+                // formatted, otherwise we would throw an error. 
+                //------------------------------------------------------------
+
+                if(options_.extractor_->riakObjectContentsCanBeParsed(value.data(), value.size())) {
```
If this condition is almost invariably be true (all data are time series data and hence satisfying the condition), then it would make sense to signal this to gcc by wrapping in it the `likely` macro.

hmmr commented on test/sut.erl:
```diff
@@ -854,8 +859,27 @@ badEncodingOptions_test() ->
+	  {range_filter, Filter},
+	  {encoding, msgpack}],
+    io:format("Filter = ~p~n", [lists:flatten([Filter])]),
+    TableName = "/Users/eml/projects/riak/riak_end_to_end_timeseries/dev/dev1/data/leveldb/" ++ Table,
```
I'm afraid make test will surely not pass on my machine :}

russelldb commented:
I don't understand _why_ these types would ever be interleaved. Wouldn't it make folds of unknown expense if TS data is not stored together?

Riak Object data is stored under a key that is `sext:encode({o, Bucket, Key})` and indexes are `sext:encode({i, Bucket, Field, Term, Key})`

Can't you ensure that TS objects have a key that ensures they are "clumped" together?

russelldb commented:
My point is, why skip over? It seems that a non TS series datum in a fold is an indication of a serious error, no?

Commit:
Added more documentation to sut.erl (unit tests).  Fixed unit tests that result in db being left open, which can cause the next test to fail

Commit:
Merging with formatBinary stuff I added earlier

Merge branch 'feature/ignore-nonmsgpack' of https://github.com/basho/eleveldb into feature/ignore-nonmsgpack

Commit:
Fixed binary tests to use proper erlang binaries.  Removed code from ErlUtil that returned pointers to non-persistent ErlNifBinary data (cut and paste error), and removed call to enif_inspect_iolist_as_binary() that causes segfault under CentOS6_x64 when used on an erlang list

hmmr commented:
+1

gordonguthrie commented:
JIRA raised to address Russell's key grouping point
https://bashoeng.atlassian.net/browse/RTS-345

gordonguthrie commented:
@borshop merge

gordonguthrie commented:
+1 069f6d4
